### ISSUE TITLE
memory leak in grpc.server

### ISSUE THREAD

**Issue by jsofri** [grpc/grpc#31509]
### What version of gRPC and what language are you using?
1.50.0

### What operating system (Linux, Windows,...) and version?
ubuntu20

### What runtime / compiler are you using (e.g. python version or version of gcc)
python3.9

### What did you do?
```python
import grpc
from concurrent import futures
tp = futures.ThreadPoolExecutor(max_workers=32)
while True:
    service=grpc.server(tp)
    service.stop(0)
```

### What did you expect to see?
same amount of memory used for process (with `top` command)

### What did you see instead?
memory keep going up. There's a memory leak in `grpc.server` implementation

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

### Anything else we should know about your project / environment?
happens in v1.47 too

**Comment by eitkin-nvidia** [comment]
Adding more information after debugging the issue on grpc **1.39.0**:

Environment
--------------
* **OS:** Ubuntu 20.04
* **Architecture:** aarch64
* **gRPC version:** grpcio-1.39.0
* **Python version:** Python 3.8.10

POC
---
```
#!/usr/bin/python3
import grpc
from concurrent import futures
import time
tp = futures.ThreadPoolExecutor(max_workers=32)
i = 0
while True:
    i += 1
    print("New Server - %d" % (i))
    service=grpc.server(tp)
    service.stop(0)
    time.sleep(0.1)
```
Changing the sleep interval doesn't change the results at all so it isn't some timing issue for Python's garbage collection.

Observations
---------------
After roughly every 80 server allocations (~50, ~130, ~210) the heap allocates a new top chunk and grows by 0x21000 (132KB) as can be viewed by running the program using strace:
```
brk(0x4314000)                          = 0x4314000
brk(0x4335000)                          = 0x4335000
brk(0x4356000)                          = 0x4356000
brk(0x4377000)                          = 0x4377000
```
The brk() behavior is aligned with the heap's growth as is explained in the following [link](https://guyinatuxedo.github.io/43-house_of_orange/house_orange_exp/index.html).

**Bottom Line:** Starting and stoping the server causes a memory leak.

**Comment by gnossen** [comment]
I think we need to come up with a more holistic way to detect issues like this. While we run our unit test suite with ASAN and MSAN, we don't have a generic framework that, for each resource type, initializes one of those resources, deinitializes the resource, and repeats, monitoring memory growth over time.

We'll look into this particular issue individually first, of course. @jsofri Do you know whether this issue has always existed? That is, have you observed this on a very old version of grpcio (such as 1.20) or do you have reason to believe that this is a regression?

**Comment by jsofri** [comment]
I didn't use older versions, so I can't tell for sure.

**Comment by sergiitk** [comment]
@asheshvidyut confirms this is still reproducible with 1.81.1.
